PDK_Abstraction/FinFET14nm_Mock_PDK/primitive.py

```diff
@@ -126,6 +143,9 @@ def _addMOSArray( self, x_cells, y_cells, pattern, connections, minvias = 2, **p
                     self._addMOS(x, y, names[0 if 0 <= ((x_cells // 2) - x) <= 1 else 1], False, **parameters)
                 else:
                     assert False, "Unknown pattern"
+                if y == y_cells-1:
+                    for x in range(x_cells):
+                        self._bodyContact(x, y, x_cells, names[0 if pattern == 0 else x_cells%2])
             self._connectDevicePins(y, connections)
         self._connectNets(x_cells, y_cells)
```

Comment:
You shouldn't need to check for `y == y_cells - 1`. Simply pull the `for x in range(x_cells)` loop outside the `for y in range(y_cells)` and `for x in range(x_cells)` loops. Basically do it at the very end with `self._connectNets(x_cells, y_cells)`